x/concentrated-liquidity/swaps.go

```diff
@@ -301,19 +318,25 @@ func (k Keeper) calcOutAmtGivenIn(ctx sdk.Context,
 			return writeCtx, sdk.Coin{}, sdk.Coin{}, sdk.Int{}, sdk.Dec{}, sdk.Dec{}, fmt.Errorf("could not convert next tick (%v) to nextSqrtPrice", nextTick)
 		}
 
+		// Charge fee
+		feeOnAmountRemainingIn := swapState.amountSpecifiedRemaining.Mul(swapFee)
+
```

Comment:
This is probably right, but im not quite understanding the flow on first pass. 

We charge a fee, then subtract that fee from the amount remaining on the swap step (this makes sense to me)

What I dont understand is why we run a seperate function `computeFeeChargePerSwapStep` to calculate the `feeChargeTotal` for the global accumulator. Why would it not just be taken from `feeOnAmountRemainingIn`?

(Once again, I am sure this is right, but on first pass its not clear why)

Reply:
Let's define what `swapState.amountSpecifiedRemaining` is.

It is the amount of token in remaining over all swap steps.

After performing the current swap step, the following cases are possible:
1.  we consumed all amount remaining
   * this is what we charge here: https://github.com/osmosis-labs/osmosis/blob/3df20182cae965f4c7d621ab947f4cee681becc1/x/concentrated-liquidity/fees.go#L292-L295
   * this essentially implies the difference between the original amount remaining and the one actually consumed. The difference between them is the fee
2. we did not consume all amount remaining
   * then we charge this: https://github.com/osmosis-labs/osmosis/blob/3df20182cae965f4c7d621ab947f4cee681becc1/x/concentrated-liquidity/fees.go#L290
   * this is the fee charged on the amount actually consumed during a swap step
3. Price impact protection makes us exit before consuming all amount remaining
* then we charge this (similar to option 2): https://github.com/osmosis-labs/osmosis/blob/3df20182cae965f4c7d621ab947f4cee681becc1/x/concentrated-liquidity/fees.go#L290
* the swap fee on the amount in actually consumed before price impact protection got trigerred

Reply:
Awesome! This is something I did not previously know clearly before! Would we be able to document this context somewhere?

Reply:
This make sense to me, thanks! If this exact writeup could get documented somewhere that would be awesome

Reply:
Docs added to `architecture.md`. The fee docs need to be refactored - quite outdated. I will come back to it after the swap in the other direction is done